### PR TITLE
[ABW-3827] Resource view Pre-Authorization states

### PR DESCRIPTION
**Pull request by danvleju-rdx**
Jira ticket: [ABW-3827](https://radixdlt.atlassian.net/browse/ABW-3827)

## Description
- Adds mapping for an open pre-authorization from ManifestSummary / StaticAnalysis
- Introduces new states to the Resource view:
    - Supports multiple types of amounts: `exact`, `atLeast`, `atMost`, `between`, `predicted`, and `unknown`. The former `ResourceAmount` becomes `ExactResourceAmount`, and the new `ResourceAmount` is an enum with different types of amounts.
    - Unknown resource. The former `ResourceBalance` becomes `KnownResourceBalance`, and `ResourceBalance` is now an enum with `known` and `unknown` variants.
- Adds support for the new amount types in resource details screens, as well as for estimated and guaranteed amounts.

### Notes
`ResourceAmount` is used in multiple places, including some where non-exact amounts don’t make sense (e.g., AccountDetails). Therefore, only exact amounts are handled in these cases.

## How to test
Use any dApp that supports Pre-Authorization.
In the screenshots below, I used:  [Test manifests.zip](https://github.com/user-attachments/files/17657523/Test.manifests.zip)

#### Amounts
|  |  |  |  |
| - | - | - | - |
|Simulator Screenshot - iPhone 16 - 2024-11-07 at 09 49 52 | Simulator Screenshot - iPhone 16 - 2024-11-07 at 09 50 12|Simulator Screenshot - iPhone 16 - 2024-11-07 at 09 50 56 | Simulator Screenshot - iPhone 16 - 2024-11-07 at 09 50 33|

| Possible dApp calls | Proofs |
| - | - |
| Simulator Screenshot - iPhone 16 - 2024-11-07 at 09 45 39 | Simulator Screenshot - iPhone 16 - 2024-11-07 at 09 46 00 |

#### Details
|  |  |  |  |
| - | - | - | - |
|  Simulator Screenshot - iPhone 16 - 2024-11-07 at 09 42 22| Simulator Screenshot - iPhone 16 - 2024-11-07 at 09 41 50 | Simulator Screenshot - iPhone 16 - 2024-11-07 at 09 31 17 | Simulator Screenshot - iPhone 16 - 2024-11-07 at 09 41 11 |

#### Guarantee
|  |  | 
| - | - |
|Simulator Screenshot - iPhone 16 - 2024-11-07 at 09 54 55 |Simulator Screenshot - iPhone 16 - 2024-11-07 at 09 54 59 |

[ABW-3827]: https://radixdlt.atlassian.net/browse/ABW-3827?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: RadixWallet/Clients/AccountPortfoliosClient/AccountPortfoliosClient+State.swift [hunk 1]
```diff
@@ -247,62 +247,90 @@ private extension AccountPortfoliosClient.AccountPortfolio {
...
+	mutating func updateFiatWorth(_ change: (ResourceAddress, ExactResourceAmount) -> FiatWorth?) {
 		mutateAll { detail in
-			let xrdRedemptionValue = detail.xrdRedemptionValue
-			detail.stakeUnitResource.mutate {
-				$0.amount.fiatWorth = change(
-					.mainnetXRD,
-					.init(
-						nominalAmount: xrdRedemptionValue,
-						fiatWorth: $0.amount.fiatWorth
+			var xrdRedemptionValue = detail.xrdRedemptionValue
```
Comment (GhenadieVP) [inline RadixWallet/Clients/AccountPortfoliosClient/AccountPortfoliosClient+State.swift:318]: Does not seem to be used.

File: RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+ComplexTypes.swift [hunk 2]
```diff
@@ -44,11 +41,35 @@ extension OnLedgerEntitiesClient {
...
+		validatorStakes: [TrackedValidatorStake] = [],
+		entities: InteractionReview.Sections.ResourcesInfo = [:],
+		resourceAssociatedDapps: InteractionReview.Sections.ResourceAssociatedDapps? = nil,
+		networkID: NetworkID,
+		defaultDepositGuarantee: Decimal192 = 1
+	) async throws -> KnownResourceBalance {
+		let resourceAddress = resource.resourceAddress
+
+		@Dependency(\.resourcesVisibilityClient) var resourcesVisibilityClient
+		let hiddenResources = try await resourcesVisibilityClient.getHidden()
```
Comment (GhenadieVP) [inline RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+ComplexTypes.swift:66]: Interesting, why was this added?

File: RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+ComplexTypes.swift [hunk 3]
```diff
@@ -72,24 +93,27 @@ extension OnLedgerEntitiesClient {
...
 		networkID: NetworkID,
 		guarantee: TransactionGuarantee?,
 		hiddenResources: [ResourceIdentifier]
-	) async throws -> ResourceBalance {
+	) async throws -> KnownResourceBalance {
 		let resourceAddress = resource.resourceAddress
 
 		if let poolContribution = poolContributions.first(where: { $0.poolUnitsResourceAddress == resourceAddress }) {
+			guard let amount = amount.exactAmount?.nominalAmount else {
+				fatalError("Not possible")
```
Comment (GhenadieVP) [inline RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+ComplexTypes.swift:115]: Add a better message as to why this is not possible.
It is not possible because the poolContributions can happen only with an enclosed manifest, where the amounts are exact.

Still, it may be worth investigating if we could have separate functions that are called when resolving the sections for manifestSummary or from executionSummary.
For the case of manifestSummary, there cannot be `poolContributions`.

File: RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+ComplexTypes.swift [hunk 4]
```diff
@@ -119,7 +143,11 @@ extension OnLedgerEntitiesClient {
 					details: .init(
 						address: poolContribution.poolAddress,
 						dAppName: resourceAssociatedDapps?[resourceAddress]?.name,
-						poolUnitResource: .init(resource: resource, amount: .init(nominalAmount: amount)),
+						poolUnitResource: .init(
+							resource: resource,
+							amount: .exact(.init(nominalAmount: amount)),
```
Comment (GhenadieVP) [inline RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+ComplexTypes.swift:148]: seeing this multiple times, maybe worth declaring a function on ResourceAmount like:
```swift
static func exact(_ nominalAmount: Decimal) -> Self
```

File: RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+ComplexTypes.swift [hunk 5]
```diff
@@ -147,38 +175,41 @@ extension OnLedgerEntitiesClient {
...
+		let worth: ResourceAmount
 		if validatorStakes.isEmpty {
 			guard let totalSupply = resource.totalSupply, totalSupply.isPositive else {
 				throw MissingPositiveTotalSupply()
 			}
 
-			worth = amount * validator.xrdVaultBalance / totalSupply
+			worth = amount.adjustedNominalAmount { $0 * validator.xrdVaultBalance / totalSupply }
 		} else {
+			guard let exactAmount = amount.exactAmount else {
```
Comment (GhenadieVP) [inline RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+ComplexTypes.swift:191]: Same as above.

File: RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+CreateEntity.swift [hunk 6]
```diff
@@ -556,12 +570,14 @@ extension OnLedgerEntitiesClient {
 		struct ResourceWithRedemptionValue: Hashable, Sendable {
 			let resource: OnLedgerEntity.Resource
 			var redemptionValue: ResourceAmount?
+			var guarantee: Decimal192?
```
Comment (GhenadieVP) [inline RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+CreateEntity.swift:573]: Could we have guarantee in ResourceAmount 🤔? As it is another representation of an amount like `exact`, `atLeast`...

File: RadixWallet/Features/AssetsFeature/Components/FungibleAssetList/Components/Details/FungibleTokenDetails+View.swift [hunk 7]
```diff
@@ -5,8 +5,9 @@ extension FungibleTokenDetails.State {
 	var detailsHeader: DetailsContainerWithHeaderViewState {
 		.init(
 			title: resource.metadata.get(\.name, prefetched: ownedFungibleResource?.metadata).map { $0 ?? L10n.Account.PoolUnits.unknownPoolUnitName },
-			amount: ownedFungibleResource?.amount.nominalAmount.formatted(),
-			currencyWorth: ownedFungibleResource?.amount.fiatWorth?.currencyFormatted(applyCustomFont: false),
+			amount: ownedFungibleResource?.amount,
+			guaranteed: ownedFungibleResource?.guarantee,
+			currencyWorth: ownedFungibleResource?.amount.exactAmount?.fiatWorth?.currencyFormatted(applyCustomFont: false), // FIXME: handle not exact amounts
```
Comment (GhenadieVP) [inline RadixWallet/Features/AssetsFeature/Components/FungibleAssetList/Components/Details/FungibleTokenDetails+View.swift:10]: I believe that we don't show fiat worth when opening details from TX Review, yet. So this is not necessarily a FIXME

File: RadixWallet/Features/AssetsFeature/Components/HelperViews/ResourceBalance/ResourceBalanceView.swift [hunk 8]
```diff
@@ -543,78 +613,229 @@ extension ResourceBalanceView {
...
+			switch appearance {
+			case .standard, .compact:
+				.trailing
+			case .large:
+				.center
+			}
+		}
+	}
+
+	struct SubAmountView: View {
```
Comment (GhenadieVP) [inline RadixWallet/Features/AssetsFeature/Components/HelperViews/ResourceBalance/ResourceBalanceView.swift:709]: What does `Sub` mean :)?

File: RadixWallet/Features/InteractionReview/Sections/Sections+ManifestSummary.swift [hunk 9]
```diff
@@ -1,6 +1,242 @@
...
+		)
+
+		let dappAddresses = summary.encounteredEntities.compactMap {
+			switch $0 {
+			case let .component(componentAddress):
+				componentAddress.isGlobal ? componentAddress.asGeneral : nil
+			case let .locker(lockerAddress):
+				lockerAddress.asGeneral
+			}
+		}
```
Comment (GhenadieVP) [inline RadixWallet/Features/InteractionReview/Sections/Sections+ManifestSummary.swift:39]: This should be common with `sections(for summary: ExecutionSummary,`, worth adding an extension to be used in both places.

File: RadixWallet/Features/InteractionReview/Sections/Sections.swift [hunk 10]
```diff
@@ -299,7 +308,7 @@ private extension InteractionReview.Sections {
 			state.destination = .nonFungibleTokenDetails(.init(
 				resourceAddress: resource.resourceAddress,
 				resourceDetails: .success(resource),
-				token: nft,
+				details: nft != nil ? .token(nft!) : nil,
```
Comment (GhenadieVP) [inline RadixWallet/Features/InteractionReview/Sections/Sections.swift:311]: Seems to be equal to be equal to:
```
nft.map(KnownResourceBalance.NonFungible.token)
```

**Review by GhenadieVP — COMMENTED** [review]
Amazing work 👏 ! Left some comments for improvements

File: RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+ComplexTypes.swift (hunk 2 above)
Comment (danvleju-rdx) [reply RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+ComplexTypes.swift:66]: Not added, just moved from the older function

File: RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+CreateEntity.swift (hunk 6 above)
Comment (danvleju-rdx) [reply RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+CreateEntity.swift:573]: Hmm, shouldn't guarantee always be an exact amount?

File: RadixWallet/Features/AssetsFeature/Components/HelperViews/ResourceBalance/ResourceBalanceView.swift (hunk 8 above)
Comment (danvleju-rdx) [reply RadixWallet/Features/AssetsFeature/Components/HelperViews/ResourceBalance/ResourceBalanceView.swift:709]: `SubAmountView` indicates it's a smaller part or subset of the main `AmountView`. I could rename it to `AmountItemView` if that’s clearer.

File: RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+CreateEntity.swift (hunk 6 above)
Comment (GhenadieVP) [reply RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+CreateEntity.swift:573]: Well all types in ResourceAmount contain the ExactAmount, like `case between(minimum: ExactResourceAmount, maximum: ExactResourceAmount)`, and the predicted amount is very similar to that and could be:
```swift
case predicted(predicted: ExactAmount, guaranteed: ExactAmount)
```
The combination of:
```swift
var redemptionValue: ResourceAmount?
var guarantee: Decimal192?
```
Sort of implies that there could be `ResourceAmount.between` + guarantee, which is not true.

Comment (danvleju-rdx) [reply RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+CreateEntity.swift:573]: Right, the `case predicted` is the right way to do it

File: RadixWallet/Clients/AccountPortfoliosClient/AccountPortfoliosClient+State.swift [hunk 11]
```diff
@@ -247,62 +247,90 @@ private extension AccountPortfoliosClient.AccountPortfolio {
...
-	mutating func updateFiatWorth(_ change: (ResourceAddress, ResourceAmount) -> FiatWorth?) {
+	mutating func updateFiatWorth(_ change: (ResourceAddress, ExactResourceAmount) -> FiatWorth?) {
 		account.fungibleResources.updateFiatWorth(change)
 		stakeUnitDetails.mutateValue { $0.updateFiatWorth(change) }
 		poolUnitDetails.mutateValue { $0.updateFiatWorth(change) }
 	}
 }
 
+extension ResourceAmount {
+	mutating func updateFiatWorth(resourceAddress: ResourceAddress, change: (ResourceAddress, ExactResourceAmount) -> FiatWorth?) {
```
Comment (Sajjon) [inline RadixWallet/Clients/AccountPortfoliosClient/AccountPortfoliosClient+State.swift:258]: I prefer line breaks for long signatures - try `CTRL + M` in Xcode :)

File: RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+ComplexTypes.swift [hunk 12]
```diff
@@ -147,38 +175,41 @@ extension OnLedgerEntitiesClient {
...
-		let worth: Decimal192
+	) async throws -> KnownResourceBalance {
+		let worth: ResourceAmount
 		if validatorStakes.isEmpty {
 			guard let totalSupply = resource.totalSupply, totalSupply.isPositive else {
 				throw MissingPositiveTotalSupply()
 			}
 
-			worth = amount * validator.xrdVaultBalance / totalSupply
+			worth = amount.adjustedNominalAmount { $0 * validator.xrdVaultBalance / totalSupply }
```
Comment (Sajjon) [inline RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+ComplexTypes.swift:188]: token arithmetic ought best be done in Sargon, would that makes sense? And with unit tests...

File: RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+ComplexTypes.swift [hunk 13]
```diff
@@ -147,38 +175,41 @@ extension OnLedgerEntitiesClient {
...
 					throw StakeUnitAddressMismatch()
 				}
 				// Distribute the worth in proportion to the amounts, if needed
-				if stake.liquidStakeUnitAmount == amount {
-					worth = stake.xrdAmount
+				if stake.liquidStakeUnitAmount == exactAmount.nominalAmount {
+					worth = .exact(.init(nominalAmount: stake.xrdAmount))
 				} else {
-					worth = (amount / stake.liquidStakeUnitAmount) * stake.xrdAmount
+					worth = .exact(.init(nominalAmount: (exactAmount.nominalAmount / stake.liquidStakeUnitAmount) * stake.xrdAmount))
```
Comment (Sajjon) [inline RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+ComplexTypes.swift]: possible to move token amount arithmetic to Sargon?

File: RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+ComplexTypes.swift [hunk 14]
```diff
@@ -258,7 +288,7 @@ extension OnLedgerEntitiesClient {
 					)
 				}
 				.map { id in
-					ResourceBalance(resource: resource, details: .nonFungible(.init(id: id, data: nil)), isHidden: false)
+					KnownResourceBalance(resource: resource, details: .nonFungible(.token(.init(id: id, data: nil))), isHidden: false)
```
Comment (Sajjon) [inline RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+ComplexTypes.swift:290]: `CTRL + M` please :)

File: RadixWallet/Core/SharedModels/Assets/OnLedgerEntity.swift [hunk 15]
```diff
@@ -555,7 +562,7 @@ extension OnLedgerEntity.OnLedgerAccount {
 		var debugDescription: String {
 			"""
 			\(validatorAddress.formatted())
-			staked: \(stakeUnitResource?.amount.nominalAmount.formatted() ?? "NONE")
+			staked: \(stakeUnitResource?.amount.debugDescription ?? "NONE")
```
Comment (Sajjon) [inline RadixWallet/Core/SharedModels/Assets/OnLedgerEntity.swift:557]: I know we are calling `.debugDescription` in a LOT of places, but it is discouraged by Apple:
https://developer.apple.com/documentation/swift/customdebugstringconvertible/debugdescription#discussion

> Calling this property directly is discouraged. Instead, convert an instance of any type to a string by using the String(reflecting:) initializer. This initializer works with any type, and uses the custom debugDescription property for types that conform to CustomDebugStringConvertible:

So perhaps we ought for new code use `String(reflecting:)` instead?

File: RadixWallet/Core/SharedModels/Assets/OnLedgerEntity.swift [hunk 16]
```diff
@@ -689,9 +696,12 @@ extension OnLedgerEntity.Resource {
 			loggerGlobal.error("Total supply is 0 for \(poolUnitResource.resource.resourceAddress.address)")
 			return nil
 		}
-		let redemptionValue = poolUnitResource.amount.nominalAmount * (amount / poolUnitTotalSupply)
+		guard case let .exact(amount) = amount else {
+			return nil
+		}
+		let redemptionValue = poolUnitResource.amount.adjustedNominalAmount { $0 * (amount.nominalAmount / poolUnitTotalSupply) }
 		let decimalPlaces = divisibility ?? Decimal192.maxDivisibility
-		let roundedRedemptionValue = redemptionValue.rounded(decimalPlaces: decimalPlaces)
```
Comment (Sajjon) [inline RadixWallet/Core/SharedModels/Assets/OnLedgerEntity.swift:694]: consider moving token amount arithmetic to Sargon?

File: RadixWallet/Features/AssetsFeature/Components/HelperViews/ResourceBalance/ResourceBalanceView.swift [hunk 17]
```diff
@@ -437,37 +460,68 @@ extension ResourceBalanceView {
 		}
 	}
 
+	struct Unknown: View {
+		var body: some View {
+			HStack(spacing: .small1) {
+				Image(.unknownDeposits)
+
+				Text("----")
```
Comment (Sajjon) [inline RadixWallet/Features/AssetsFeature/Components/HelperViews/ResourceBalance/ResourceBalanceView.swift:468]: perhaps `"----"` ought to be "localized"? so that Matt can change it?

File: RadixWallet/Features/TransactionReviewFeature/TransactionReview.swift [hunk 18]
```diff
@@ -584,22 +584,33 @@ extension ResourceBalance {
...
+					var known = knownResourceBalance
+					known.details = .liquidStakeUnit(liquidStakeUnit)
+					self = .known(known)
+				case var .poolUnit(poolUnit):
+					poolUnit.guarantee = newValue
+					var known = knownResourceBalance
+					known.details = .poolUnit(poolUnit)
+					self = .known(known)
+				case .nonFungible, .stakeClaimNFT:
+					return
```
Comment (Sajjon) [inline RadixWallet/Features/TransactionReviewFeature/TransactionReview.swift:611]: feels wrong to NOOP the `newValue` in case current value was `.nonFungible, . stakeClaimNFT, . unknown`, I Know we used to do it (before your changes) for `.nonFungible, . stakeClaimNFT`. But it looks wrong.

Can we at least add an `assertionFailure`?

**Review by CyonAlexRDX — APPROVED** [review]
Bravo!

Just prefer having new models in Sargon.

And Token artithmetic we have had loads of bugs with, and it is very important to get right, I prefer having that in Sargon.

But we will get there... can't do it all in one go, just unfortunate to further add/extend token amount arithmetic in host

File: RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+ComplexTypes.swift (hunk 3 above)
Comment (danvleju-rdx) [reply RadixWallet/Clients/OnLedgerEntitiesClient/Helpers/OnLedgerEntitiesClient+ComplexTypes.swift]: Did the separation [here](https://github.com/radixdlt/babylon-wallet-ios/pull/1377/commits/9ad60a00f7de3a6ad41dae88b9ab35a90c0ccd2c)

File: RadixWallet/Features/AssetsFeature/Components/HelperViews/ResourceBalance/ResourceBalanceView.swift (hunk 17 above)
Comment (danvleju-rdx) [reply RadixWallet/Features/AssetsFeature/Components/HelperViews/ResourceBalance/ResourceBalanceView.swift:468]: This is a placeholder used for unknown deposits. It shouldn’t need to be changed, but if it does, it will likely require UI changes as well.

**Review by GhenadieVP — APPROVED** [review]
